### PR TITLE
Feature/friendship v2

### PR DESCRIPTION
**Pull request by bini-i**
## The following modifications are added to the project
- create_mutual_friendship method added to the friendship model.
- uses after_update callback to add a symmetric record to build a mutual friendship.
- update friends method in the user model.
- friends method looks up friends of a user in one query.

### CODE REVIEW

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

File: app/models/user.rb [hunk 1]
```diff
@@ -34,9 +37,15 @@ def reject_request(user)
...
-    friends.concat(invitations_i_sent.map { |friendship| friendship.invitee if friendship.accepted })
-    friends.compact
+    # friends = invitations_i_got.map { |friendship| friendship.inviter if friendship.accepted }
+    # friends.concat(invitations_i_sent.map { |friendship| friendship.invitee if friendship.accepted })
+    # friends.compact
+
+    # SELECT * FROM users WHERE id IN (
+    #   SELECT friendships.inviter_id From friendships WHERE accepted = true    [1,2,5]
+    # )
+    # inviters_id = u2.invitations_i_got.where(accepted: true).pluck(:inviter_id)
```
Comment (descholar-ceo) [inline app/models/user.rb]: Remove the commented codes

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @bini-i and @yocosaka  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is great now and it is really nice, congratulations on the good work you have done so far.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] Please remove the commented codes wherever they are in your code
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Binyam Hailemeskel & Yoko Saka** 
Good job with the improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
Just one last detail to improve before you can go ahead:

- [x] In your README, the link to the screenshot is broken
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by Srjnnnn — APPROVED** [review]
## Status: Approved :partying_face: :tada: 
Hello team! I've checked your app and it works great. :muscle: I'd like to mention some things below so you can address them later.

- [x] You've used an older version of Sass which is deprecated. You can use Dart Sass instead of it.


- [x]  Running local server gives me some warnings. Probably the version of your Rails app. There are some deprecated syntax there. You can think about it and you'll fix those later.



- [x]  Not all of your tests have passed at my side. But it's because of the models created already.  I've checked the tests and give you permission to move on.
- [x] Your live demo link is directing me to the same repo. You can create a live demo and link it. I am approving this PR since the requirements are not asking for a live demo at this stage.

Again, it is a great job. Keep up the good work. :100: 
Congratulations and good luck on your next projects.
_Eyüp Sercan UYGUR_

**Comment by yokoworks** [comment]
@eypsrcnuygr , Thanks a lot for your great suggestion.

### About Rails version and other gems' versions
I'd like to suggest you update your Rails version to solve some errors that have occurred in your environment.

I think, as you see, the console says, this project is using Rails 2.7.2, but your Rails is in version 2.7.0.
104623635-9e465700-56a3-11eb-907e-a7bacae79936

We started this project by cloning the repository that was prepared by Microverse.
So we didn't specify the version of SCSS and other gems that were already set by Microverse.

Actually, in my local environment, the same kind of errors occurred before.
So I updated my Rails version to 2.7.2, and now these errors disappeared.

### About rspec
I just tried to create the complete same User objects for just running tests, that were defined for the RSpec test inside the method in user_rspec, then run RSpec again, and now all the tests have been passed correctly.

I guess if we define the object inside spec files like below, these objects don't affect the actual data in the project's database.
```
describe 'testing instance methods' do
    let(:user1) { User.create(id: 1, name: 'user1', email: 'user1@gmail.com', password: '123456') }
    let(:user2) { User.create(id: 2, name: 'user2', email: 'user2@gmail.com', password: '123456') }
...
```
So I don't know why all the tests were not passed in your environment.


Anyway, thank you so much for your great comments and suggestions.
We'll keep it brush up.

**Pull request merged**